File: 1696-jump-game-vi/1696-jump-game-vi.cpp
class Solution {
public:
    int maxResult(vector<int>& nums, int k) {
        int n = nums.size();
        priority_queue<pair<int, int>> maxHeap;
        vector<int> dp(n, 0);
        
        maxHeap.push({nums[0], 0});
        dp[0] = nums[0];
        
        for(int i = 1; i < n; ++i) {
            while(maxHeap.top().second < i - k)
                maxHeap.pop();
            
            dp[i] = nums[i] + dp[maxHeap.top().second];
            
            maxHeap.push({dp[i], i});            
        }
        return dp[n - 1];
    }
};

File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        maxHeap = []
        dp = [0] * len(nums)
        
        dp[0] = nums[0]
        heappush(maxHeap, (-nums[0], 0))
        
        for i in range(1, len(nums)):
            while maxHeap[0][1] < i - k:
                heappop(maxHeap)
            
            dp[i] = nums[i] + dp[maxHeap[0][1]]
            
            heappush(maxHeap, (-dp[i], i))
        
        return dp[-1]